File: backend/src/main/scala/domains/blog/model/BlogSummary.scala
package domains.blog.model

import domains.auth.model.UserIdentity
import domains.problem.model.{ProblemSlug, ProblemTitle}
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

import java.time.Instant
import scala.util.Try

final case class BlogSummary(
  id: BlogId,
  title: BlogTitle,
  content: BlogContent,
  author: UserIdentity,
  visibility: BlogVisibility,
  blogType: BlogType,
  problemSlug: Option[ProblemSlug],
  problemTitle: Option[ProblemTitle],
  score: Int,
  viewerVote: Option[BlogVote],
  createdAt: Instant,
  updatedAt: Instant
)

object BlogSummary:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap { value =>
    Try(Instant.parse(value)).toEither.left.map(_.getMessage)
  }

  given Encoder[BlogSummary] = deriveEncoder[BlogSummary]
  given Decoder[BlogSummary] = deriveDecoder[BlogSummary]

File: frontend/src/features/blog/model/BlogSummary.ts
import type { UserIdentity } from '@/features/auth/domain/auth'
import type { BlogContent } from '@/features/blog/model/BlogContent'
import type { BlogId } from '@/features/blog/model/BlogId'
import type { BlogTitle } from '@/features/blog/model/BlogTitle'
import type { BlogType } from '@/features/blog/model/BlogType'
import type { BlogVisibility } from '@/features/blog/model/BlogVisibility'
import type { BlogVote } from '@/features/blog/model/BlogVote'
import type { ProblemSlug, ProblemTitle } from '@/features/problem/domain/problem'

export type BlogSummary = {
  id: BlogId
  title: BlogTitle
  content: BlogContent
  author: UserIdentity
  visibility: BlogVisibility
  blogType: BlogType
  problemSlug: ProblemSlug | null
  problemTitle: ProblemTitle | null
  score: number
  viewerVote: BlogVote | null
  createdAt: string
  updatedAt: string
}
